crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1093 @@
+                    };
+                    Some(TwoQubitBasisDecomposer::new_inner(
+                        pi_2_basis.to_string(),
+                        StandardGate::CXGate.matrix(&[]).unwrap().view(),
+                        fidelity,
+                        basis_1q,
+                        Some(true),
+                    )?)
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+
+            let decomposer = xx_decomposer.call1((
+                &basis_2q_fidelity_dict,
+                PyString::new_bound(py, basis_1q),
+                &embodiments_dict,
+                pi2_decomposer,
+            ))?;
+            let decomposer_gate = decomposer
+                .getattr(intern!(py, "gate"))?
+                .extract::<NormalOperation>()?;
+
+            decomposers.push(DecomposerElement {
+                decomposer: DecomposerType::XXDecomposer(decomposer.into()),
+                gate: decomposer_gate,
+            });
+        }
+    }
+    Ok(Some(decomposers))
+}
+
+fn preferred_direction(
+    decomposer: &DecomposerElement,
+    ref_qubits: &[PhysicalQubit; 2],
+    natural_direction: Option<bool>,
+    coupling_edges: &Bound<'_, PyList>,
+    target: &Target,
+) -> PyResult<Option<bool>> {
+    // Returns:
+    // * true if gate qubits are in the hardware-native direction
+    // * false if gate qubits must be flipped to match hardware-native direction∂
+    let qubits: [PhysicalQubit; 2] = *ref_qubits;
+    let mut reverse_qubits: [PhysicalQubit; 2] = qubits;
+    reverse_qubits.reverse();
+
+    let compute_cost =
+        |lengths: bool, q_tuple: [PhysicalQubit; 2], in_cost: f64| -> PyResult<f64> {
+            let cost = match target.qargs_for_operation_name(decomposer.gate.operation.name()) {
+                Ok(_) => match target[decomposer.gate.operation.name()].get(Some(
+                    &q_tuple
+                        .into_iter()
+                        .collect::<SmallVec<[PhysicalQubit; 2]>>(),
+                )) {
+                    Some(Some(_props)) => {
+                        if lengths {
+                            _props.duration.unwrap_or(in_cost)
+                        } else {
+                            _props.error.unwrap_or(in_cost)
+                        }
+                    }
+                    _ => in_cost,
+                },
+                Err(_) => in_cost,
+            };
+            Ok(cost)
+        };
+
+    let preferred_direction = match natural_direction {
+        Some(false) => None,
+        _ => {
+            // None or Some(true)
+            let mut edge_set = HashSet::new();
+            for item in coupling_edges.iter() {
+                if let Ok(tuple) = item.extract::<(usize, usize)>() {
+                    edge_set.insert(tuple);
+                } else if let Ok(inner_list) = item.extract::<&PyList>() {
+                    if inner_list.len() == 2 {
```

Comment:
Use downcast for this line
```suggestion
                } else if let Ok(inner_list) = item.downcast::<PyList>() {
```
Or if you'd rather analyze it fully in Rust, try extracting into rust_native types. Seems like it would be an array of type `[usize; 2]` or a tuple `(usize, uize)`.
```suggestion
                } else if let Ok(inner_list) = item.extract::<[usize; 2]>() {
```
or 

```suggestion
                } else if let Ok(inner_list) = item.downcast::<(usize, usize)>() {
```

Reply:
I think that this was a bit of an overkill anyways, and the first if already extracts to a tuple of usize and that should probably be enough, so I just removed the else. Thanks for pointing this out!